File: ADOPCION_I/02_documentacion/ejemplo_despues/con_documentacion.py
"""
Código DESPUÉS: Con documentación generada por Copilot
El mismo código pero ahora es fácil de entender y mantener
"""

def process_payment(amount, currency, method):
    """
    Procesa un pago calculando fees y retornando información de la transacción.
    
    Esta función calcula el fee de procesamiento (2.9% + $0.30), determina
    el estado de la transacción según el método de pago, y retorna un objeto
    con todos los detalles.
    
    Args:
        amount (float): Monto del pago en la moneda especificada. Debe ser > 0.
        currency (str): Código de moneda ISO 4217 (ej: 'USD', 'EUR', 'MXN').
        method (str): Método de pago. Valores válidos: 'card', 'bank', 'cash'.
    
    Returns:
        dict or None: Diccionario con información de la transacción conteniendo:
            - amount (float): Monto original
            - fee (float): Fee de procesamiento calculado
            - total (float): Monto total incluyendo fee
            - currency (str): Moneda de la transacción
            - method (str): Método de pago utilizado
            - status (str): Estado de la transacción ('approved', 'pending_verification')
        
        Retorna None si el monto es <= 0.
    
    Examples:
        >>> process_payment(100.00, 'USD', 'card')
        {
            'amount': 100.0,
            'fee': 3.20,
            'total': 103.20,
            'currency': 'USD',
            'method': 'card',
            'status': 'approved'
        }
        
        >>> process_payment(-50, 'USD', 'card')
        None
    
    Note:
        - Los pagos con tarjeta son aprobados inmediatamente
        - Los pagos bancarios requieren verificación
        - El fee es fijo: 2.9% + $0.30
    """
    if amount <= 0:
        return None
    
    fee = amount * 0.029 + 0.30
    total = amount + fee
    
    transaction = {
        'amount': amount,
        'fee': fee,
        'total': total,
        'currency': currency,
        'method': method,
        'status': 'pending'
    }
    
    if method == 'card':
        transaction['status'] = 'approved'
    elif method == 'bank':
        transaction['status'] = 'pending_verification'
    
    return transaction


def validate_user_input(data):
    """
    Valida datos de entrada del usuario según reglas de negocio.
    
    Verifica que el diccionario de datos contenga email válido, contraseña
    segura y edad apropiada (si está presente).
    
    Args:
        data (dict): Diccionario con datos del usuario a validar. Puede contener:
            - email (str): Dirección de email (requerido)
            - password (str): Contraseña (requerido, min 8 caracteres)
            - age (int, optional): Edad del usuario
    
    Returns:
        tuple: Tupla de (is_valid, errors) donde:
            - is_valid (bool): True si todos los datos son válidos
            - errors (list): Lista de strings con nombres de campos inválidos
    
    Examples:
        >>> validate_user_input({'email': 'user@example.com', 'password': 'SecurePass123'})
        (True, [])
        
        >>> validate_user_input({'email': 'invalid', 'password': '123'})
        (False, ['email', 'password'])
        
        >>> validate_user_input({'email': 'user@example.com', 'password': 'SecurePass123', 'age': 15})
        (False, ['age'])
    
    Validation Rules:
        - Email debe contener '@'
        - Password debe tener al menos 8 caracteres
        - Age (si existe) debe ser >= 18
    """
    errors = []
    
    if 'email' not in data or '@' not in data['email']:
        errors.append('email')
    
    if 'password' not in data or len(data['password']) < 8:
        errors.append('password')
    
    if 'age' in data and data['age'] < 18:
        errors.append('age')
    
    return len(errors) == 0, errors


class UserManager:
    """
    Administrador de usuarios con caché para optimizar consultas a base de datos.
    
    Esta clase maneja operaciones CRUD de usuarios implementando un sistema
    de caché simple para reducir consultas repetidas a la base de datos.
    
    Attributes:
        db: Conexión a la base de datos
        cache (dict): Diccionario de caché {user_id: user_data}
    
    Examples:
        >>> db = DatabaseConnection()
        >>> manager = UserManager(db)
        >>> user = manager.get_user(123)
        >>> manager.update_user(123, {'name': 'John Doe'})
    
    Note:
        El caché se invalida automáticamente cuando se actualiza un usuario.
    """
    
    def __init__(self, db_connection):
        """
        Inicializa el UserManager con una conexión a base de datos.
        
        Args:
            db_connection: Objeto de conexión a la base de datos que implementa
                          métodos query() y execute().
        """
        self.db = db_connection
        self.cache = {}
    
    def get_user(self, user_id, use_cache=True):
        """
        Obtiene un usuario por ID, opcionalmente usando caché.
        
        Args:
            user_id (int): ID único del usuario a buscar.
            use_cache (bool, optional): Si True, usa caché para lecturas repetidas.
                                       Default True.
        
        Returns:
            dict or None: Diccionario con datos del usuario si existe, None en caso contrario.
                         Estructura típica: {'id': int, 'name': str, 'email': str}
        
        Examples:
            >>> manager.get_user(123)  # Primera llamada: consulta DB y cachea
            {'id': 123, 'name': 'John', 'email': 'john@example.com'}
            
            >>> manager.get_user(123)  # Segunda llamada: retorna desde caché
            {'id': 123, 'name': 'John', 'email': 'john@example.com'}
            
            >>> manager.get_user(123, use_cache=False)  # Fuerza consulta a DB
            {'id': 123, 'name': 'John', 'email': 'john@example.com'}
        
        Performance:
            - Con caché: ~0.001ms
            - Sin caché: ~10-50ms (dependiendo de la DB)
        """
        if use_cache and user_id in self.cache:
            return self.cache[user_id]
        
        user = self.db.query('SELECT * FROM users WHERE id = ?', [user_id])
        
        if user and use_cache:
            self.cache[user_id] = user
        
        return user
    
    def update_user(self, user_id, data):
        """
        Actualiza datos de un usuario e invalida su entrada en caché.
        
        Args:
            user_id (int): ID del usuario a actualizar.
            data (dict): Diccionario con campos a actualizar. Puede contener:
                        - name (str, optional): Nuevo nombre
                        - email (str, optional): Nuevo email
        
        Returns:
            bool: True si la actualización fue exitosa, False en caso contrario.
        
        Examples:
            >>> manager.update_user(123, {'name': 'Jane Doe'})
            True
            
            >>> manager.update_user(123, {'name': 'Jane', 'email': 'jane@example.com'})
            True
        
        Note:
            - Solo actualiza campos presentes en 'data'
            - Invalida automáticamente el caché para este usuario
            - Los cambios son permanentes en la base de datos
        
        Raises:
            DatabaseError: Si hay error en la consulta SQL
        """
        result = self.db.execute(
            'UPDATE users SET name=?, email=? WHERE id=?',
            [data.get('name'), data.get('email'), user_id]
        )
        
        # Invalidar caché
        if user_id in self.cache:
            del self.cache[user_id]
        
        return result


if __name__ == "__main__":
    # Ejemplos de uso con documentación clara
    print("=== Procesamiento de Pagos ===")
    transaction = process_payment(100, 'USD', 'card')
    print(f"Transacción: {transaction}")
    
    print("\n=== Validación ===")
    valid, errors = validate_user_input({
        'email': 'test@example.com',
        'password': 'SecurePass123'
    })
    print(f"Válido: {valid}, Errores: {errors}")

File: ADOPCION_I/02_documentacion/ejemplo_despues/con_documentacion.js
/**
 * Código DESPUÉS: Con documentación completa usando JSDoc
 * Este código es fácil de entender gracias a la documentación detallada
 */

/**
 * Procesa un pago calculando las comisiones y generando una transacción.
 * 
 * Esta función calcula la comisión del procesador de pagos (2.9% + $0.30),
 * suma el total y crea un objeto de transacción con el estado apropiado
 * según el método de pago utilizado.
 * 
 * @param {number} amount - Monto del pago en la moneda especificada. Debe ser mayor a 0.
 * @param {string} currency - Código de moneda (ej: 'USD', 'EUR', 'MXN').
 * @param {'card'|'bank'} method - Método de pago ('card' para tarjeta o 'bank' para banco).
 * @returns {Object|null} Objeto de transacción con amount, fee, total, currency, method y status, o null si el monto es inválido.
 * 
 * @example
 * // Pago con tarjeta aprobado automáticamente
 * processPayment(100, 'USD', 'card');
 * // Returns: { amount: 100, fee: 3.20, total: 103.20, currency: 'USD', method: 'card', status: 'approved' }
 * 
 * @example
 * // Pago con banco requiere verificación
 * processPayment(500, 'EUR', 'bank');
 * // Returns: { amount: 500, fee: 14.80, total: 514.80, currency: 'EUR', method: 'bank', status: 'pending_verification' }
 * 
 * @example
 * // Monto inválido retorna null
 * processPayment(-50, 'USD', 'card');
 * // Returns: null
 */
function processPayment(amount, currency, method) {
    if (amount <= 0) {
        return null;
    }
    
    const fee = amount * 0.029 + 0.30;
    const total = amount + fee;
    
    const transaction = {
        amount: amount,
        fee: fee,
        total: total,
        currency: currency,
        method: method,
        status: 'pending'
    };
    
    if (method === 'card') {
        transaction.status = 'approved';
    } else if (method === 'bank') {
        transaction.status = 'pending_verification';
    }
    
    return transaction;
}

/**
 * Valida los datos de entrada del usuario según reglas de negocio.
 * 
 * Verifica que el email sea válido, la contraseña cumpla requisitos mínimos
 * y la edad sea apropiada para el servicio.
 * 
 * @param {Object} data - Datos del usuario a validar.
 * @param {string} data.email - Email del usuario. Debe contener '@'.
 * @param {string} data.password - Contraseña del usuario. Debe tener al menos 8 caracteres.
 * @param {number} [data.age] - Edad del usuario (opcional). Si se proporciona, debe ser >= 18.
 * @returns {{isValid: boolean, errors: string[]}} Objeto con isValid (true si no hay errores) y array de campos con errores.
 * 
 * @example
 * // Validación exitosa
 * validateUserInput({ email: 'user@example.com', password: 'securePass123', age: 25 });
 * // Returns: { isValid: true, errors: [] }
 * 
 * @example
 * // Validación con errores
 * validateUserInput({ email: 'invalid', password: '123', age: 16 });
 * // Returns: { isValid: false, errors: ['email', 'password', 'age'] }
 */
function validateUserInput(data) {
    const errors = [];
    
    if (!data.email || !data.email.includes('@')) {
        errors.push('email');
    }
    
    if (!data.password || data.password.length < 8) {
        errors.push('password');
    }
    
    if (data.age && data.age < 18) {
        errors.push('age');
    }
    
    return {
        isValid: errors.length === 0,
        errors: errors
    };
}

/**
 * Gestor de usuarios con cache integrado para optimizar consultas a la base de datos.
 * 
 * Esta clase proporciona métodos para consultar y actualizar información de usuarios,
 * con un sistema de cache en memoria para reducir la carga en la base de datos.
 * 
 * @class
 * @example
 * const db = new DatabaseConnection();
 * const userManager = new UserManager(db);
 * const user = userManager.getUser(123);
 */
class UserManager {
    /**
     * Crea una instancia del gestor de usuarios.
     * 
     * @param {Object} dbConnection - Objeto de conexión a la base de datos con métodos query() y execute().
     */
    constructor(dbConnection) {
        /** @type {Object} Conexión a la base de datos */
        this.db = dbConnection;
        
        /** @type {Object.<number, Object>} Cache de usuarios en memoria */
        this.cache = {};
    }
    
    /**
     * Obtiene un usuario de la base de datos o del cache.
     * 
     * Si el cache está habilitado y el usuario existe en cache, lo retorna directamente.
     * Si no, consulta la base de datos y opcionalmente guarda el resultado en cache.
     * 
     * @param {number} userId - ID del usuario a buscar.
     * @param {boolean} [useCache=true] - Si es true, usa y actualiza el cache. Si es false, siempre consulta la BD.
     * @returns {Object|null} Objeto con los datos del usuario o null si no existe.
     * 
     * @example
     * // Obtener usuario usando cache
     * const user = userManager.getUser(123);
     * 
     * @example
     * // Obtener usuario sin usar cache (consulta directa a BD)
     * const freshUser = userManager.getUser(123, false);
     */
    getUser(userId, useCache = true) {
        if (useCache && this.cache[userId]) {
            return this.cache[userId];
        }
        
        const user = this.db.query('SELECT * FROM users WHERE id = ?', [userId]);
        
        if (user && useCache) {
            this.cache[userId] = user;
        }
        
        return user;
    }
    
    /**
     * Actualiza los datos de un usuario en la base de datos.
     * 
     * Ejecuta una actualización en la base de datos y elimina el usuario del cache
     * para asegurar que futuras consultas obtengan los datos actualizados.
     * 
     * @param {number} userId - ID del usuario a actualizar.
     * @param {Object} data - Datos a actualizar.
     * @param {string} data.name - Nuevo nombre del usuario.
     * @param {string} data.email - Nuevo email del usuario.
     * @returns {Object} Resultado de la operación de actualización.
     * 
     * @example
     * const result = userManager.updateUser(123, { 
     *   name: 'Juan Pérez', 
     *   email: 'juan@example.com' 
     * });
     */
    updateUser(userId, data) {
        const result = this.db.execute(
            'UPDATE users SET name=?, email=? WHERE id=?',
            [data.name, data.email, userId]
        );
        
        // Invalidar cache después de actualizar
        if (this.cache[userId]) {
            delete this.cache[userId];
        }
        
        return result;
    }
}

module.exports = { processPayment, validateUserInput, UserManager };
